pkg/route/secret/secret_monitor.go

```diff
@@ -0,0 +1,211 @@
+package secret
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	routev1 "github.com/openshift/api/route/v1"
+	"github.com/openshift/library-go/pkg/route/monitor"
+)
+
+type listObjectFunc func(string, metav1.ListOptions) (runtime.Object, error)
+type watchObjectFunc func(string, metav1.ListOptions) (watch.Interface, error)
+
+// Monitor manages Kubernetes secrets. This includes retrieving
+// secrets or registering/unregistering them via Routes.
+type Monitor interface {
```

Comment:
I think we need something like

```go
type SecretEventHandlerRegistration interface{
  ResourceEventHandlerRegistration

  UID() string
}

type SecretMonitor interface{
  AddEventHandler(namespace, name string, handler ResourceEventHandler) (SecretEventHandlerRegistration, error)
  
  RemoveEventHandler(SecretEventHandlerRegistration) error
}
```

This will allow an eventHandler for every route (overlapping secrets).  It's not perfectly efficient, but it is easy.  And handler passed in could wire back to a route to be queued in a different controller.  Need a link to your calling code to see how to wire that back in.

We might also need a way to get a list of all known eventHandlers to diff against the current state (re-list after an outage), but if the caller correctly handles deletes we may not need it.  It is a potential spot for a leak.

Reply:
See the 1.27 SharedInformer interface for where I got inspiration and how can you wire it up through a sharedinformer versus a regular informer.

Reply:
So I understand your suggestion as, we need an abstraction between the secret watching logic and And the caller (the component that interfaces with Routes) in form of (do correct me if I understood it incorrectly)
```go
type SecretMonitor interface{
  AddEventHandler(namespace, name string, handler ResourceEventHandler) (SecretEventHandlerRegistration, error)
  
  RemoveEventHandler(SecretEventHandlerRegistration) error
}
```

> See the 1.27 SharedInformer interface for where I got inspiration and how can you wire it up through a sharedinformer versus a regular informer.

Are you also suggesting I replace the use of the regular informer in the implementation with a sharedinformer?